src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketSenderPool.cs

```diff
@@ -0,0 +1,55 @@
+using System;
+using System.Collections.Concurrent;
+using System.IO.Pipelines;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    internal class SocketSenderPool : IDisposable
+    {
+        private const int MaxQueueSize = 1024; // REVIEW: Is this good enough?
```

Comment:
Did you look to see how many were created with 5000 websockets?

Reply:
yes, 8 on my machine (8 cores).

Reply:
1024 seems high to me. Is this the maximum number of sends occurring at the same time?

Reply:
> 1024 seems high to me. Is this the maximum number of sends occurring at the same time?

What number do you like? A multiple of the number of cores?

Reply:
Also, this is a maximum, it doesn't mean we start with this many or that we'll have this many. It means we won't got past this many.

Reply:
I understand, but once you get up to 1024, there isn't a way to reduce the amount pooled.

Maybe something like 128 would be better.

Reply:
> A multiple of the number of cores?

16 * cores?

Reply:
Right now we allocate one of these queues per IOQueue, we could pick a conservative number like 128. The 1024 was copied from what we used to do in libuv for a similar pool (the WriteQueue) https://github.com/dotnet/aspnetcore/blob/90c45d94ac5f2127063e841964f87188fbda05bc/src/Servers/Kestrel/Transport.Libuv/src/Internal/WriteReqPool.cs#L10.

Reply:
Maybe we should make it 32 so it always fits in a single `ConcurrentQueueSegment`. We'd have to be strict about never going over that in `Return(SocketSender)`.

Reply:
We don't want to be strict. I'll pick whatever number ya'll want though as long as we don't introduce locking into this code path.

Reply:
You can be strict without locking. You can use `Interlocked.Increment/Decrement`. `ConcurrentQueue.Enqueue` is already doing it's own `Interlocked.Increment` at a minimum.

Reply:
OK I've complicated the code.

Reply:
Also, there's no way that I can think of to make this accurate without locking. Review the code and let me know.

Reply:
```suggestion
        // Similar to https://github.com/dotnet/aspnetcore/blob/90c45d94ac5f2127063e841964f87188fbda05bc/src/Servers/Kestrel/Transport.Libuv/src/Internal/WriteReqPool.cs#L10.
        private const int MaxQueueSize = 1024;
```